### ISSUE TITLE
[REST] Return all metadata by default

### ISSUE THREAD

**Issue by spacemanspiff2007** [openhab/openhab-core#3300]
Currently the metadata selector for the  `/items/{itemname}` has to be used to return the item metadata.
To get all metadata one has to use a regular expression, e.g. `.+`.
However not all users know regular expressions and it's not clear for all users that metadata can be requested this way.

I propose a change in behavior:
By default the whole metadata for the item is returned. 
The metadata selector can be used to reduce the returned metadata.
That way it would be a backwards compatible change.

---

This is a follow up issue for #3298

**Comment by florian-h05** [comment]
If there returning all metadata does not lead to performance drawbacks: +1

**Comment by spacemanspiff2007** [comment]
@J-N-K since you're the master of metadata I'd like to hear your opinion. I think from a usability perspective it absolutely makes sense. Do you have any performance concerns?

**Comment by J-N-K** [comment]
Not sure. The code for fetching metadata is quite complex due to the applied filters, this can be easily improved if we simply want "everything". 

With the current code fetching all items (360 item) from my production system without metadata takes around 350 ms (107 kB) and fetching all items with `.+` takes around 400 ms (123 kB). This is just a quick test made with the API Explorer from a remote machine, so I'm not sure how much is really related to backend processing or network transfer. 

Anyway, it seems that getting all metadata takes about 15% more time in this scenario. It's worth noting that the range without metadata was 337 to 482 ms and with metadata 389 to 437 ms (I tried ten times each). So maybe the "real life" difference is small. Maybe you can try that on your systems, too.

**Comment by spacemanspiff2007** [comment]
I've written a short script that makes 50 requests and measures the time against my production system.
(140k without metadata, 150k with metadata)

````
Without metadata
Avg: 313.7ms, best: 282.0ms, worst: 438.0ms
Quantile: 50% between 297.0ms and 313.0ms

With metadata
Avg: 320.6ms, best: 297.0ms, worst: 344.0ms
Quantile: 50% between 312.0ms and 328.0ms

Difference:
Avg: +2.2%, best: +5.1%, worst: -27.3%
Quantile:+4.8% and +4.6%
````

On average the request with the metadata filter took 2.2% longer, however the 50% of the requests took ~5% longer.

I run it a couple of times and the avg response time has a +-20% jitter and the worst response time was 300% of the previous worst response time.
So it seems to takes longer, but it's not the limiting factor and the response time is absolutely not deterministic anyway.


The mainUI already requests the items with a metadata filter so for the majority of users nothing will change.
If we change the default behavior experience will improve.

**Comment by ccutrer** [comment]
I currently have 2,099 items, 1,316 have some sort of metadata (though 283 of them are the semantics auto-generated semantics namespace, and 68 are stateDescriptions). The majority of my metadata is configuration for homekit and alexa. I do use it for a few other things, but not extensively.

`/rest/items` is 681kB.
10 iterations of `time curl http://localhost:8080/rest/items > /dev/null`:
Avg: 282.7ms, best: 235ms, worst: 331ms

`/rest/items?metadata=.*` is 824kB
10 iterations of `time curl http://localhost:8080/rest/items?metadata=.* > /dev/null`
Avg: 331.5ms, best: 295ms, worst: 357ms

Difference:
Avg: +11.9%, best: +25.5%, worst: 7.9%.

**Comment by rkoshak** [comment]
My system is relatively modest when it comes to Items. I have only 375 Items. I'm not sure the best way to figure out how many have metadata without counting by hand. Most of my metadata are state descriptions, default X widgets, and various stuff used by my rules. If it helps, my JSONDB metadata file is 169kB.

`/rest/items` is 1332kB (how do I have an order of magnitude fewer Items but twice the size as @ccutrer?)
10 iterations:
Avg: 226.8 ms, best: 198, worst: 259

`/rest/items?metadata=.*` is 1404 kB
10 iterations:
Avg: 235.4ms, best: 207ms, worst: 207ms

Difference: +3.7%, best: +11.3%, worst: +4.3%

I'm on a well apportioned VM running on a server class machine (an old one but still pretty zippy for what I need it to do). I ran the curl commands from a different machine but it's a wired 1 GB connection so network shouldn't impact the measurements too much.

One thing to note here though is I bet most of us are either not defining default widgets at all, or we are defining custom widgets and then applying them to our Items. But if one is defining complex widgets directly on their Items, the metadata can become quite large. 

JustinG (that's his handle on the forum, I don't know what he goes by on GitHub so I sent him a PM) uses Item metadata extensively. He might be an extreme case worth running the test.

**Comment by JustinGeorgi** [comment]
Happy to contribute, fwiw.  My system is only medium sized: 890 items.  If you count only standard metadata and my custom namespaces I'd estimate that 25% of those have metadata, once you factor in the autogenerated semantic metadata I'd say that probably jumps to about 85 - 90%.

`/rest/items` is 2393kB
10 iterations:
Avg: 125.5 ms, best: 121, worst: 129

`/rest/items?metadata=.*` is 2497 kB
10 iterations:
Avg: 154.2 ms, best: 146, worst: 179

Difference: +22.7%, best: +20.7%, worst: +38.8%

I've got OH running in docker on a linux box (that, honestly, is starting to show its age).

**Comment by spacemanspiff2007** [comment]
> It looks like you have very little metadata (7%, even for me it's only 15% and I don't have much metadata set)

I have ~200 items with metadata but it's mostly just a key/value pair.

---

Can you describe where you see performance issue because if the user has to make one additional request (e.g. to get metadata for one item the "performance gain" is lost). And metadata is more and more used to define how an item behaves so I don't see how the information without is still useful.

And again: 
There is no "real" Performance impact because MainUI already uses the metadata filter to request items together with metadata.
So things will not get worse for the average user, but usability for the RestAPI user will go up.
And obviously usability is bad because somebody created a PR to get the metadata namespaces (and it got merged) because it was not clear that `/items` and `/items/{itemname}` could return metadata.

It's also inconsistent:
All other parameters reduce the returned items (if passed) but metadata will increase the returned values (if passed).
I'd even say since metadata is more or less part of the item object that it's even a violation of the REST principle.


If performance is super important on can always
- use the `fields` parameter
- Still use an metadata selector (e.g. `asdf`)
- If you're still unsure we could also use a magic value for metadata (e.g. `EXCLUDE`) to not return it

**Comment by ccutrer** [comment]
So thinking about this more from the lens of someone that has highly optimized web APIs for very large websites in the past, I've had a few thoughts:

 * When on your local network, a 10-20% increase in size and latency of this API isn't going to matter much.
 * I'm much more concerned about the _size_ of the response going up 10% as I am about speed. Specifically I think about the Alexa integration, where it has to pull a bunch of data through openHABCloud, and then process it all, within a time limit specified by Amazon (the maintainers over there already had to implement some optimizations in order to be able to process the size of my configuration in time). Something that might be reasonable here is compressing the response over HTTP (possibly intelligently, only if the client is not local).
 * WRT speed, IIRC the MetadataRegistry stores everything with a composite key of (item name, metadata namespace). Even the new helper that allows you to get all metadata for an item without knowing namespace just iterates all entries. That means that for a REST request where we're iterating all items, and then getting all metadata for each item, it's n^2. There are definitely opportunities for optimizing this - either having the resource handler just grab all the metadata in a massive blob, so that it can iterate it once building a hash table with items as keys, or in the MetadataRegistry using a sorted map instead of a hash map, so that given an item name it's n log n to find the first entry for an item without knowing any namespaces, and then just iterate from there, vastly speeding up the getAllNamespaces call. Or just maintaining an additional caching data structure of HashMap<item name, Set<namespaces> >. Several options here.
  
Tl;dr: I agree with @spacemanspiff2007 that it would be much easier and intuitive to have all metadata included by default. Should probably still be coordinated with major API consumers, _especially_ cloud based ones (openhabCloud, Alexa integration, Google Assistant integration?, mobile apps? come to mind, but there might be more) to ensure that they won't suddenly take an unexpected performance hit, especially wrt transfer size. And then if it's still seen as a hit, spend some time optimizing metadata lookup.

**Comment by rkoshak** [comment]
> I'm much more concerned about the size of the response going up 10% as I am about speed.

Good point. Any increase we make here is going to be multiplied by X where X is the number of users logged in through myopenhab.org. Can those handle a 10-20% increase in this REST API call? Probably but it's worth asking. 

The calls made by Alexa and Google Assistant integration though I would expect to be limited to just the metadata they care about by using the filter. they shouldn't be making a default call in the first place so I'm not sure the change discussed here *would* impact them, beyond changing it if necessary so they do in fact only query for their own metadata and not everything (if they are doing something else).

> IIRC the MetadataRegistry stores everything with a composite key of (item name, metadata namespace).

Correct but in the other order, namespace goes first.

**Comment by spacemanspiff2007** [comment]
> Can those handle a 10-20% increase in this REST API call? Probably but it's worth asking.

If they log in and use MainUI the metadata is already requested today.

If size is a concern then imho it would be better to use middleware in the openHAB webserver that can gzip the json response.
I looked at it and am very surprised that it's not already the case.
Typically there is a threshold e.g. everything > X kb is gzipped.
For me this would results in 90% less traffic.

**Comment by ccutrer** [comment]
Getting very tangential here, but brotli should be preferred over gzip if the client supports it (especially since a lot of our installs are on modest CPU devices). 

As for MainUI requesting all metadata - is that true for pages, or only for configuration? I only use MainUI for configuration so far, so I'm not super familiar with it.

**Comment by J-N-K** [comment]
Jetty unfortunately only supports gzip at the moment.

**Comment by rkoshak** [comment]
> If size is a concern then imho it would be better to use middleware in the openHAB webserver that can gzip the json response.

I don't know that it's a concern or not. It's a potential problem that should be validated or shown not to be one. That's all.    Do we know that right now *all* calls to the items REST API endpoint from MainUI always ask for all the metadata every time? If so then yes, there's no actual change from that perspective. But if there are some calls to the endpoint where they don't ask for any metadata now there will be an increase.  We need to at least assess what the impact will actually be. Maybe it's no big deal but I don't think we can just assert that.

> As for MainUI requesting all metadata - is that true for pages

Pages need to know the default widget metadata so I would expect it to either pull all the metadata or at least the well known metadata is needs to get those default widgets. So I don't expect those calls to be changed by this. It's only those pages where the items end point is hit but no metadata is requested now that there would be an impact. Maybe the Settings -> Items page? Developer Sidebar? Settings -> Model? Settings -> Rules/Scripts/Schedule? Developer Tools -> Widget Builder? I can't think of anywhere else that endpoint might be hit where the metadata isn't needed.

**Comment by spacemanspiff2007** [comment]
> Getting very tangential here, but brotli should be preferred over gzip if the client supports it (especially since a lot of our installs are on modest CPU devices).

> Jetty unfortunately only supports gzip at the moment: [eclipse/jetty.project#2553](https://github.com/eclipse/jetty.project/issues/2553)

It's not a problem at all because the client sends the accepted encodings to the server.
The server then picks the corresponding compression algorithm or none.

So if the client sends "gzip" the server can pick "gzip".
If the client sends "brotli" the server will pick no compression because currently there is no brotli support.
So it's no problem supporting multiple algorithms.

Screenshot when accessing the items endpoint:

![grafik](https://user-images.githubusercontent.com/10754716/212389282-747bb87d-4e18-4d1a-af92-1484e76dee41.png)

> Maybe the Settings -> Items page? 

Does also request metadata.

**Comment by JustinGeorgi** [comment]
> > Maybe the Settings -> Items page?
> 
> Does also request metadata.

The main items list does not call the metadata:

```javascript
    this.$oh.api.get('/rest/items').then((data) => {
      this.items = data.sort((a, b) => {
        const labelA = a.label || a.name
        const labelB = b.label || b.name
        return labelA.localeCompare(labelB)
      })
```

There are a handful of smaller places where the endpoint is also called with no metadata: the item picker popup, the main settings menu (just to get the number of items in the system), the item autofill popups in various editors are the ones I can think of off the top of my head.

**Comment by spacemanspiff2007** [comment]
> The main items list does not call the metadata:

Settings -> Items does call metadata

![grafik](https://user-images.githubusercontent.com/10754716/212449702-c98aea10-de65-4464-9fc2-7b752bf2ca4f.png)

---

I'm not arguing there are no places where this is used. There should indeed be a follow up issue for the UIs where the maintainer is made aware of the change.

But even with no code change (for any API consumer)
- It doesn't matter if it takes 0% - 10% longer because the API jitter is up to 200% longer
- The worst avg number reported here was 50ms longer which can't be noticed
- The change is backwards compatible so this won't break anything
- If desired metadata can be explicitly excluded through a request parameter

**Comment by JustinGeorgi** [comment]
> Settings -> Items does call metadata
> 
> ![grafik](https://user-images.githubusercontent.com/10754716/212449702-c98aea10-de65-4464-9fc2-7b752bf2ca4f.png)

Yep, you're right, my bad.  I had clicked on the old list file not the new vlist which, of course, calls semantics...

> I'm not arguing there are no places where this is used. There should indeed be a follow up issue for the UIs where the maintainer is made aware of the change.

Didn't mean to imply that just because those other uses exist that's a knock against this proposal, I'm in favor.  I was just enumerating the few places I could think of where an explicit exclusion would be preferential.

**Comment by lolodomo** [comment]
Can be closed by merged PR #3311 ?

**Comment by spacemanspiff2007** [comment]
If somebody wants to one can still optimize the logic.
Currently the regex is used to match all entries, which means that that the usage of the regex is not necessary.

However I think this made it easier for non tech savvy people so I am happy.
Thanks.
